### DISCUSSION TITLE
HHT-alpha time integration vs external forces (Neumann BCs)

### DISCUSSION THREAD

**Discussion by rcontinuum** [idaholab/moose#23209] [Q&A Modules: Solid mechanics] [answered]
Hello, 

I'm trying to use in (old) TensorMechanics the HHT-alpha time integrator and I want to apply a time dependent external mechanical force ("dead load") given as a piecewise linear function.

I do that with 

```
[Functions]
  [load]
    type = PiecewiseLinear
    x = '0.0 0.25 0.75 1.0'    # time 
    y = '0.0 2.0 -2.0  0.0'    # value   
  []
[]

[BCs]
 [pull_x]
    type = FunctionNeumannBC
    boundary = right
    variable = disp_x
    function = load
  []
[]
```
In the documentation of the HHT-alpha method, see Section "Hilber-Hughes-Taylor time integration" in
https://mooseframework.inl.gov/moose/modules/tensor_mechanics/Dynamics.html
the external force $F_{ext}$ (here `FunctionNeumannBC`) should be evaluated at time $t_\alpha = t+(1+\alpha)\Delta t$ . $t$ would be $t_{old}$.
But I presume, I may be wrong, that the Function `type = PiecewiseLinear` is always evaluated at the current time which would be $t_{new} = t+\Delta t$ and not at the (generalised) midpoint time $t_\alpha$ given by the parameter $\alpha \in [-1/3, 0]$. 

Am I right or wrong about this? Any chance to get the parameter $\alpha$ into `type = PiecewiseLinear`? Or in any other predefined function, so that the function is evaluated at the midpoint time?

Cheers

PS:  Other mechanical external forces $F_{ext}$ as, e.g.,  `Pressure` (follower force) 
https://mooseframework.inl.gov/moose/source/bcs/Pressure.html
and `Gravity` 
https://mooseframework.inl.gov/source/kernels/Gravity.html
have an additional parameter `(hht_)alpha` to evaluate  $F_{ext}$  at the correct midpoint time. Or maybe the evaluation is done by linear interpolation using $\alpha$ and $F_{ext}(t_{old})$ and $F_{ext}(t_{new})$. I didn't check the implementation.

PPS: I was able to get $t_{new}, \Delta t$ and $\alpha$ into a `ParsedFunction`, so I could use that to define a `PiecewiseLinear`
function which is evaluated at the correct midpoint time. That looks a bit "clunky" ... but seems to work.

**Comment by hugary1995** [comment, marked as answer]
You are correct in that `FunctionNeumannBC` does not honor HHT-alpha time integration.

For now, there are two workarounds
- Use `ParsedFunction` as you discovered;
- Create your own class that derives from `FunctionNeumannBC`, in which you'll handle the time integration by querying the function at time `_t + (1 + _alpha ) * _dt`.

This is a design issue, so feel free to open a feature request ticket. I don't guarantee an ETA for the fix though. You'll have to consult with the MOOSE team. As always, pull requests are welcome.
